### PR TITLE
fix remaining create test

### PR DESCRIPTION
**Pull request by invocamanman**
- Fix multiple JMPN
- Fix nonce length
- Fix static call
- Refactor on calls and creates, first substract gas then create context and edge cases of them

### CODE REVIEW

File: main/opcodes.zkasm [hunk 1]
```diff
@@ -2330,13 +2329,12 @@ opSaveTopicsLoop:
...
 ; // TODO check revert create
 opCREATE:
-
     SP - 3              :JMPN(stackUnderflow)
     $ => A              :MLOAD(isStaticCall)
     0 - A               :JMPN(invalidStaticTx)
     SP - 1 => SP
-    $ => A              :MLOAD(SP--)            ;value
-    A                   :MSTORE(valueCall)
+    $ => D              :MLOAD(SP--)            ;value
```
Comment (ignasirv) [inline main/opcodes.zkasm:2336]: Why changing register to D?

**Review by ignasirv — APPROVED** [review]

**Review by laisolizq — APPROVED** [review]

File: main/opcodes.zkasm (hunk 1 above)
Comment (invocamanman) [reply main/opcodes.zkasm:2336]: bc i can sue it later without to loading it again ^^

File: main/load-tx-rlp-utils.zkasm [hunk 2]
```diff
@@ -12,7 +12,7 @@ addBatchHashData:
 getTxBytes:
         $ => A                          :MLOAD(batchL2DataLength)
         $ => B                          :MLOAD(batchL2DataParsed)
-        A - B - C - D                   :JMPN(invalidTxRLP)
+        A - B - C - D                   :JMPN(invalidTxRLP) ; TODO CHECK JUMPN only checks 64 bits
```
Comment (krlosMata) [inline main/load-tx-rlp-utils.zkasm:15]:
```suggestion
        A - B - C - D                   :JMPN(invalidTxRLP) ; TODO CHECK JUMPN only checks 32 bits
```

File: main/load-tx-rlp-utils.zkasm [hunk 3]
```diff
@@ -21,7 +21,7 @@ getTxBytes:
 addHashTx:
         $ => A                          :MLOAD(batchL2DataLength)
         $ => B                          :MLOAD(batchL2DataParsed)
-        A - B - C - D                   :JMPN(invalidTxRLP)
+        A - B - C - D                   :JMPN(invalidTxRLP) ; TODO CHECK JUMPN only checks 64 bits
```
Comment (krlosMata) [inline main/load-tx-rlp-utils.zkasm:24]:
```suggestion
        A - B - C - D                   :JMPN(invalidTxRLP) ; TODO CHECK JUMPN only checks 32 bits
```

File: main/load-tx-rlp-utils.zkasm (hunk 2 above)
Comment (krlosMata) [reply main/load-tx-rlp-utils.zkasm:15]: We could check that `A` register is below 32 bits and then perform the `JMPN` instruction.
`A` contains `batchL2DataLength` which comes from the input, the other regusters are intenrally computed.
I think this is probably not needed since we are checking that `batchL2DataLength` is equal to `batchHashPos` in https://github.com/0xPolygonHermez/zkevm-rom/blob/main/main/main.zkasm#L142
and the `ASSERT` instructions checks all `opX` fields elements.
Therefore, I will just remove the comment

File: main/load-tx-rlp-utils.zkasm (hunk 3 above)
Comment (krlosMata) [reply main/load-tx-rlp-utils.zkasm:24]: same as previous comment

File: main/load-tx-rlp.zkasm [hunk 4]
```diff
@@ -323,7 +324,7 @@ invalidTxRLP:
         $ => E                          :MLOAD(batchHashDataId)
 
 appendTxs:
-        B - C - 32                      :JMPN(finalAppendTxs)
+        B - C - 32                      :JMPN(finalAppendTxs) ; TODO check, JUMPN only chekcs first 64 bits
```
Comment (krlosMata) [inline main/load-tx-rlp.zkasm:327]:
```suggestion
        B - C - 32                      :JMPN(finalAppendTxs) ; TODO check, JUMPN only chekcs first 32 bits
```

File: main/load-tx-rlp.zkasm [hunk 5]
```diff
@@ -333,7 +334,7 @@ appendTxs:
 
 finalAppendTxs:
         B - C => D
-        D - 1                           :JMPN(endAppendTxs)
+        D - 1                           :JMPN(endAppendTxs) ; TODO check, JUMPN only chekcs first 64 bits
```
Comment (krlosMata) [inline main/load-tx-rlp.zkasm:337]:
```suggestion
        D - 1                           :JMPN(endAppendTxs) ; TODO check, JUMPN only chekcs first 32 bits
```

File: main/load-tx-rlp.zkasm (hunk 4 above)
Comment (krlosMata) [reply main/load-tx-rlp.zkasm:327]: I would remove the comments because the same reason as `load-tx-rlp-utils.zkasm`

File: main/load-tx-rlp.zkasm (hunk 5 above)
Comment (krlosMata) [reply main/load-tx-rlp.zkasm:337]: I would remove the comments because the same reason as `load-tx-rlp-utils.zkasm`

File: main/main.zkasm [hunk 6]
```diff
@@ -94,7 +94,7 @@ txLoopRLP:
 
         $ => A          :MLOAD(batchL2DataLength)
         $ => C          :MLOAD(batchL2DataParsed)
-        C - A           :JMPN(loadTx_rlp)
+        C - A           :JMPN(loadTx_rlp) ; TODO CHECK JUMPN only checks 64 bits
```
Comment (krlosMata) [inline main/main.zkasm:97]: I would remove the comments because the same reason as `load-tx-rlp-utils.zkasm`

File: main/main.zkasm [hunk 7]
```diff
@@ -108,7 +108,7 @@ endCheckRLP:
 txLoop:
         $ => A          :MLOAD(pendingTxs)
         A-1 => A        :MSTORE(pendingTxs)
-        A               :JMPN(processTxsEnd)
+        A               :JMPN(processTxsEnd) ; TODO CHECK JUMPN only checks 64 bits
```
Comment (krlosMata) [inline main/main.zkasm:111]: `pendingTxs` is computed internally.
This check will give us a maximum number of 2³² transactions to process in one batch which I think it is save xD

File: main/process-tx.zkasm [hunk 8]
```diff
@@ -313,7 +314,7 @@ deploy:
         A => E
         ; check address to deploy is empty
                                        :CALL(isEmptyAccount)
```
Comment (krlosMata) [inline main/process-tx.zkasm:316]: We should be able to deploy on an address that has some balance.
Here we are checking that the new address is totally empty meaning that `balance == nonce == code == 0x`.
I think in case of deployment we should only check that `nonce == code == 0x`

File: main/opcodes.zkasm [hunk 9]
```diff
@@ -710,7 +710,7 @@ opCALLDATACOPY:
     E               :MSTORE(lastMemOffset)
     C               :MSTORE(lastMemLength)
     GAS - 3 => GAS  :JMPN(outOfGas)
-    GAS - ${3*((C+31)/32)} => GAS    :JMPN(outOfGas)
+    GAS - ${3*((C+31)/32)} => GAS    :JMPN(outOfGas) ; Arith
```
Comment (krlosMata) [inline main/opcodes.zkasm:713]: Add `ARITH` instruction to compute safety the gas cost

**Review by krlosMata — CHANGES_REQUESTED** [review]

**Review by krlosMata — APPROVED** [review]
As discused AFK with @invocamanman, all the changes commented in this PR would be addressed in a following PR just to split changes in two PR for the sake of clarity
